compiler/modules/front/codeGen.bal

```diff
@@ -1897,40 +1903,69 @@ function typedOperand(bir:Operand operand) returns TypedOperand? {
 }
 
 function operandPairOrderType(CodeGenContext cx, bir:Operand left, bir:Operand right) returns bir:OrderType? {
-    if operandIsNil(left) {
-        return promoteToOptOrderType(operandOrderType(cx, right));
-    }
-    if operandIsNil(right) {
-        return promoteToOptOrderType(operandOrderType(cx, left));
-    }
     bir:OrderType? lot = operandOrderType(cx, left);
     bir:OrderType? rot = operandOrderType(cx, right);
+    if lot == () || rot == () {
+        return ();
+    }
     if lot == rot {
         return lot;
     }
-    lot = promoteToOptOrderType(lot);
-    rot = promoteToOptOrderType(rot);
+    [lot, rot] = promoteOrderTypePair(lot, rot);
     if lot == rot {
         return lot;
     }
     return ();
 }
 
-function promoteToOptOrderType(bir:OrderType? ot) returns bir:OrderType? {
-    if ot is bir:UniformOrderType {
-        return { opt: ot };
+function promoteOrderTypePair(bir:OrderType? lot, bir:OrderType? rot) returns [bir:OrderType?, bir:OrderType?] {
+    // Both are bir:nilOrderType case handled before
+    if lot == bir:nilOrderType {
+        bir:OrderType? rOpt = promoteOrderType(rot);
+        return [rOpt, rOpt];
     }
-    else {
-        return ot;
+    if rot == bir:nilOrderType {
+        bir:OrderType? lOpt = promoteOrderType(lot);
+        return [lOpt, lOpt];
     }
+    match [lot, rot] {
+        // Below casts are valid since lot.opt/rot.opt is bir:NoOrderType case handled above
+        [{ opt: var l }, { opt: var r }] => {
+            var [lOpt, rOpt] = promoteOrderTypePair(<bir:OrderType>l, <bir:OrderType>r);
+            return [promoteOrderType(lOpt), promoteOrderType(rOpt)];
+        }
+        [{ opt: var l }, _] => {
+            var [lOpt, rOpt] = promoteOrderTypePair(<bir:OrderType>l, rot);
+            return [promoteOrderType(lOpt), promoteOrderType(rOpt)];
+        }
+        [_, { opt: var r }] => {
+            var [lOpt, rOpt] = promoteOrderTypePair(lot, <bir:OrderType>r);
+            return [promoteOrderType(lOpt), promoteOrderType(rOpt)];
+        }
+    }
+
+    if lot == bir:emptyTupleOrderType {
+        return [rot, rot];
+    }
+    if rot == bir:emptyTupleOrderType {
+        return [lot, lot];
+    }
+    if lot is bir:ArrayOrderType && rot is bir:ArrayOrderType {
+        // Can cast since memberOrderType is bir:NoOrderType case handled above
+        var [l, r] = promoteOrderTypePair(<bir:OrderType>lot.memberOrderType, <bir:OrderType>rot.memberOrderType);
+        bir:ArrayOrderType? lArrOrdTy = l == () ? () : { memberOrderType: l } ;
+        bir:ArrayOrderType? rArrOrdTy = r == () ? () : { memberOrderType: r } ;
+        return [ lArrOrdTy, rArrOrdTy ];
+    }
+    return [lot, rot];
 }
 
-function operandIsNil(bir:Operand operand) returns boolean {
-    if operand is bir:Register {
-        return operand.semType === t:NIL;
+function promoteOrderType(bir:OrderType? ot) returns bir:OrderType? {
```

Comment:
Why isn't this returning an OptOrderType?

Seems like what this is doing is going from an OrderType to an OptOrderType, so maybe name should something like `optionalizeOrderType`.